Go-E Charger: Einfaches caching unnötiger API Statusabfragen innerhalb 2000ms

Waldmensch1 opened the pull request:
Habe es so umgebaut, wie es in der Cloud Abfrage bereits gemacht ist. Der Go-Echarger wird nur noch 1x abgefragt anstatt 8x innerhalb einer Sekunde. Der cache Wert wird im Api Konstruktor gesetzt und kann später von dort aus konfigurierbar gemacht werden, falls nötig. 2000ms scheint mir ein guter Wert, die 8 Anfragen sind nach 350ms bei mir abgearbeitet. Kommt wahrscheinlich auf die Komplexität des restlichen Setup an. Zur Vorsicht habe ich 2 TRACE Einträge zugefügt, um den Mechanismus beobachten zu können. Die können vielleicht später entfernt werden.

Ich bin nicht in Go zuhause, bitte seht es mir nach, wenn irgendwo ein style Fehler drin ist.

Commit:
Go-eCharger, caching multiple API requests within 2 seconds

Commit:
Go-eCharger, simplify cache mechanism and include Api V1

Waldmensch1 committed:
Merge branch 'evcc-io:master' into master

andig commented:
Der Cache müsste mindestens invalidiert werden wenn irgendetwas umgeschaltet wird. Ich bin momentan noch nicht überzeugt, dass wir das ohne Not umsetzen wollen.

Waldmensch1 commented:
> Der Cache müsste mindestens invalidiert werden wenn irgendetwas umgeschaltet wird. Ich bin momentan noch nicht überzeugt, dass wir das ohne Not umsetzen wollen

nach meinen Beobachtungen vergehen zwischen 2 (Aktions-)Intervallen immer 10 Sekunden. Zwischen den Intervallen kommen nur mqtt subscriptions rein, auf die aber nicht reagiert wird. Nach meiner Beobachtung (allerdings hab ich nur den Go-E) wird im Fall von Umschaltungen "Update" benutzt, was eine eigene Response hat. Diese ist unabhängig vom Statusabruf und würde auch nicht gecached. Der Go-E macht innerhalb 1-2 Sekunden überhaupt nichts, was irgendwie auswertbar wäre. Man könnte den Intervall theoretisch auch auf 1000ms setzen. Das würde die 8 Abfragen, zumindest auf meinem System, auch noch einfangen. Beim Cloud Abruf ist der cache Wert irgendwie flexibel angelegt. Da habe ich aber die Spur nicht so wirklich rückverfolgen können, wo der tatsächlich als Wert gesetzt wird. Da fehlt mir die Kenntnis des kompletten Designs. Ich habe leider zu viel anderes Sachen auf dem Tisch, um mich tief einzuarbeiten, zumal ich bisher auch keine Berührungspunkte mit Go hatte.

Rein interessehalber, was ist bei dem Build check schiefgegangen? Sieht aus wie Probleme mit whitespaces.

andig commented:
Deinem Quelltext fehlt die Behandlung mit `gofmt`.

Commit:
Go-e Charger, fmt file

Commit:
Go-eCharger, invalidate cache on update charger

Waldmensch1 committed:
Merge branch 'evcc-io:master' into master

Commit:
Merge branch 'master' of https://github.com/Waldmensch1/evcc

Waldmensch1 commented:
Invalidierung ist jetzt mit drin, falls der Charger ge-updated wird. Selbe Methodik wie in der Cloud Api bereits umgesetzt ist

andig commented on charger/go-e/api.go:
```diff
@@ -32,8 +32,12 @@ type API interface {
+	status  Response
+	log     *util.Logger
+	updated time.Time
+	cache   int64
```
time.Duration

andig commented on charger/go-e/api.go:
```diff
@@ -78,20 +84,32 @@ func (c *LocalAPI) response(partial string, res interface{}) error {
-	}
+func (c *LocalAPI) Status() (res Response, err error) {
+	if time.Since(c.updated).Milliseconds() > c.cache {
+		c.log.TRACE.Println("Native Response, cache age:", time.Since(c.updated).Milliseconds(), "ms /", c.cache, "ms")
```
raus

Waldmensch1 replied:
Du meinst, die Logeinträge raus? Den einen, oder alle? Wenn alle, dann kann der ganze else Zweig weg und das log kann aus der struct auch komplett raus. Das habe ich nur reingeholt um diesen Codeteil zu loggen.

andig replied:
Alle wenn dus nicht mehr zum debuggen brauchst

Waldmensch1 replied:
Okay, Type gefixt und log ist komplett raus.

Waldmensch1 committed:
Merge branch 'evcc-io:master' into master

Commit:
Go-eCharger, fix types, remove debug log

andig commented on charger/go-e/api.go:
```diff
@@ -45,6 +48,7 @@ func NewLocal(log *util.Logger, uri string) *LocalAPI {
 	api := &LocalAPI{
 		Helper: request.NewHelper(log),
 		uri:    uri,
+		cache:  time.Duration(2000) * time.Millisecond,
```
2 * time.Second

andig commented on charger/go-e/api.go:
```diff
@@ -78,20 +82,28 @@ func (c *LocalAPI) response(partial string, res interface{}) error {
-	}
+func (c *LocalAPI) Status() (res Response, err error) {
+	if time.Since(c.updated) > c.cache {
+
```
leerzeile raus

andig replied:
ping

Waldmensch1 replied:
Also jetzt wirst Du aber pingelig 😂
Moment

Waldmensch1 replied:
Done

andig replied:
Yo, is mein Job. Ich muss es ja warten ;)

Waldmensch1 commented on charger/go-e/api.go:
```diff
@@ -78,20 +82,28 @@ func (c *LocalAPI) response(partial string, res interface{}) error {
+	if time.Since(c.updated) > c.cache {
+
+		// fork API V2
+		if c.v2 {
```
Sonst müsste ich ja 2x die Zeit abfragen. Einmal für V2 und unten (originalcode) für V1. Okay, man könnte die Zeitabfrage auch ganz außen rumlegen. So fand ich es aber optisch schöner und selbsterklärender

Waldmensch1 replied:
Ich habe grad nochmal in die Doku geschaut, was würdest Du davon halten, wenn wir den cache Parameter einfach auch bis in die lokale API durchreichen? Dann könnten wir das cachen default ausgeschaltet lassen und auf User Wunsch einschalten. Der Parameter ist sowieso da, also wird die Config nicht verkompliziert. Wäre der perfekte Kompromiss, oder?


doku für Cloud Config 
```
chargers:
  - name: my_charger
    type: go-e
    token: 4711c # go-e cloud API token
    cache: 10s # go-e cloud API cache duration
```

andig replied:
ich wollts grad vorschlagen!

Waldmensch1 replied:
Habe aktuell den Tisch voll, versuche es heute Nachmittag zu committen

Waldmensch1 replied:
Okay, ist drin. Bei den go-e Templates passt aber so einiges nicht. Zum Beispiel kann man auch keinen Cloud Token konfigurieren. Das müsste nochmal überarbeitet werden. Ich habe die Templates auch noch nicht ganz verstanden, daher habe ich da lieber nichts angefasst. Eine native go-e konfiguration funktioniert einwandfrei, mit token bzw. uri und cache. In dem api_test.go sind 2 Aufrufe zu NewLocal(). Ich glaube zwar nicht, das der Code aktiv benutzt wird (habe alles nach Funktionsaufrufen TestLocalV1() / TestLocalV2() abgesucht und nichts gefunden), aber ich habe da mal den zusätzlichen Parameter mit eingefügt, damit es keinen Fehler reißt, falls jemand den code doch ausführt.

Konfiguration fehlerfrei getestet, auch mit fehlendem cache parameter. Dieser ist dann 0 und entsprechend geht jeder call durch zur Wallbox.
```
chargers:
- type: go-e
  uri: 192.168.178.147  # IP-Adresse oder Hostname
  name: go-e
  cache: 1s
```

Waldmensch1 committed:
Merge branch 'evcc-io:master' into master

Commit:
Go-eCharger, use cache config parameter from cloud config in local Api as well

andig commented on charger/go-e/api_test.go:
```diff
@@ -37,7 +38,7 @@ func TestLocalV1(t *testing.T) {
 
 	// h.expect("/api/status?filter=alw")
-	local := NewLocal(util.NewLogger("foo"), srv.URL)
+	local := NewLocal(util.NewLogger("foo"), srv.URL, time.Duration(0))
```
0

andig commented on charger/go-e/api_test.go:
```diff
@@ -54,7 +55,7 @@ func TestLocalV2(t *testing.T) {
 
 	h.expect("/api/status?filter=alw")
-	local := NewLocal(util.NewLogger("foo"), srv.URL)
+	local := NewLocal(util.NewLogger("foo"), srv.URL, time.Duration(0))
```
dito

Waldmensch1 replied:
Done, den import von "time" entsprechend auch wieder raus

Commit:
Go-eCharger, removed time reference in api_test.go

Commit:
Go-eCharger, remove whitspaces

andig commented on charger/go-e/api.go:
```diff
@@ -78,20 +82,25 @@ func (c *LocalAPI) response(partial string, res interface{}) error {
-		err := c.response("status?filter=alw,car,eto,nrg,wh,trx,cards", &res)
-		return res, err
+func (c *LocalAPI) Status() (res Response, err error) {
+	if time.Since(c.updated) > c.cache {
```
Verstehe ich immer noch nicht. Warum musste localapi das vorher nicht zu unterscheiden und du musst das jetzt?

Waldmensch1 replied:
Hä? Im originalen Code wird oben nach v2 geprüft und Responsev2 geliefert und die Funktion innerhalb des If verlassen. Wenn V2 nicht zutrifft wurde der untere Code ausgeführt. Es wurde im Alten Code unterschieden und im neuen Code wird auch unterschieden. Warum es 2 unterschiedliche Responses gibt, weiß ich nicht. Das habe ich so nicht gebaut, das war schon vorher so.
Ich habe jetzt keinen Bock mehr, hier über Trivialitäten und leerzeilen zu diskutieren, für solche Spielchen ist mir echt die Zeit zu schade. Nimm es oder rejecte es. Ich nutze dann halt meinen Fork. Ich wollte nur unterstützen, aber sowas habe ich noch in keinem Open Source Projekt erlebt

andig commented:
> Hä? Im originalen Code wird oben nach v2 geprüft und Responsev2 geliefert und die Funktion innerhalb des If verlassen. Wenn V2 nicht zutrifft wurde der untere Code ausgeführt. Es wurde im Alten Code unterschieden und im neuen Code wird auch unterschieden. Warum es 2 unterschiedliche Responses gibt, weiß ich nicht. Das habe ich so nicht gebaut, das war schon vorher so.
Ich habe jetzt keinen Bock mehr, hier über Trivialitäten und leerzeilen zu diskutieren, für solche Spielchen ist mir echt die Zeit zu schade. Nimm es oder rejecte es. Ich nutze dann halt meinen Fork. Ich wollte nur unterstützen, aber sowas habe ich noch in keinem Open Source Projekt erlebt

Ich muss den Code noch warten wenn Du weg bist. Deshalb muss ich ihn verstehen. Kann gut sein, dass sich die Zusammenhänge im Github Diff UI nicht richtig erschlossen haben. 

Da Dir es aber anscheinend zuviel ist mich dabei zu unterstützen den PR zu verstehen und hier uaf eine persönliche Ebene kommen willst won‘tfix.

Waldmensch1 commented:
Originalcode
```
// Status reads a v1/v2 api response
func (c *LocalAPI) Status() (Response, error) {
	if c.v2 {
		res := new(StatusResponse2)
		err := c.response("status?filter=alw,car,eto,nrg,wh,trx,cards", &res)
		return res, err
	}

	res := new(StatusResponse)
	err := c.response("status", &res)
	return res, err
}
```

Mein Code

```
// Status reads a v1/v2 api response
func (c *LocalAPI) Status() (res Response, err error) {
	if time.Since(c.updated) > c.cache {
		if c.v2 {
			c.status = new(StatusResponse2)
			err = c.response("status?filter=alw,car,eto,nrg,wh,trx,cards", &c.status)
		} else {
			c.status = new(StatusResponse)
			err = c.response("status", &c.status)
		}
		if err == nil {
			c.updated = time.Now()
		}
	}
	return c.status, err
}
```

Wo ist der Unterschied? Beide Male wird nach v2 unterschieden. Das ist doch eindeutig. Wie gesagt, mir ist es egal. Lass mal den Thread nochmal Revue passieren und überlege, ob das so okay ist. Erst verlangst Du Typ Konformität bezüglich time.Duration. Dann halte ich mich dran in api_test.go - da ist es dann falsch und ich soll 0 nehmen. Dann stört eine Leerzeile. Wenn das der Umgang in diesem Projekt ist, wird es wohl nicht viele Leute geben, die auf so einen Spießrutenlauf Lust haben. Das hat auch nicht mit persönlicher Ebene zu tun. Ich bin im Beruf Senior Developer und mache normalerweise peer reviews. Wenn ich einen solchen Stil an den Tag legen würde, kämen wir nie voran.

Wahrscheinlich liest Du das eh nicht, aber es ist wenigstens dokumentiert für die Nachwelt

Waldmensch1 committed:
Merge branch 'evcc-io:master' into master

andig commented:
> Erst verlangst Du Typ Konformität bezüglich time.Duration. Dann halte ich mich dran in api_test.go - da ist es dann falsch und ich soll 0 nehmen 

Hör Dir mal bitte selbst zu. Ich „verlange“ gar nichts. An den Code in diesem Repository habe ich aber Qualitätsansprüche. Dir musst Du nicht teilen. 

Da Go bei untypisierten integers selbst die Konvertierung in den Zieltyp macht war die time.Duration an der Stelle schlicht nicht notwendig.

Waldmensch1 committed:
Merge branch 'evcc-io:master' into master

andig commented:
@Waldmensch1 wollen wirs nochmal miteinander versuchen und den PR mergen?

Waldmensch1 commented:
> @Waldmensch1 wollen wir's nochmal miteinander versuchen und den PR mergen?

Ich habe damit kein Problem damit, ansonsten hätte ich ihn zurückgezogen. Der PR ist okay und läuft, so wie er ist seit ein paar Tagen bei mir. Ich wüsste nicht, was noch dran geändert werden müsste. Ob man die V1 nun im else Zweig macht oder einfach am Ende macht, wenn die Funktion nicht vorher verlassen wurde, macht für mich keinen großen Unterschied.

Ich würde lediglich das **fwv** hier
`err = c.response("status?filter=alw,car,eto,nrg,wh,trx,cards", &c.status)`
noch anfügen, weil die Firmware Version bei StatusResponse2 m.E. nicht abgefragt wird, in den types2.go jedoch enthalten ist. In den letzten Tagen hat man gesehen, dass die FW Version zukünftig eventuell doch eine Rolle spielen könnte (Stichwort amx Parameter). Dann würde man vermutlich bei deren Abfrage und vorliegender V2 ein nil oder nullref kriegen. Das habe ich aber nicht getestet, weil ich evcc aktuell aus meinem Fork direkt auf dem Raspi gebaut und wieder dort laufen hab.

Waldmensch1 committed:
Merge branch 'evcc-io:master' into master

andig commented:
Ich hab den fwv jetzt rausgelassen. Die FW-Anforderung ist dokumentiert. Danke für den PR!

andig commented:
@Waldmensch1 wollen wir noch einen 1s default setzen? Der könnte dann allerdings vom Anwender nicht mehr mit 0 überschrieben werden.

Waldmensch1 commented:
Naja, das hab ich mir nicht getraut, weil es das Standardverhalten ändert. So wie es jetzt ist, muss sich der Nutzer bewusst für das Caching entscheiden, wenn der Parameter nicht benutzt wird bleibt das herkömmliche Verhalten. Andererseits sehe ich kein technisches Problem für ein 1s default, da der go-E eh die gleichen Daten liefert. Der braucht immer ein paar Sekunden, bis der in Wallung kommt.

OffTopic:
Am liebsten hätte ich, wenn der Go-E über kommunizieren würde, weil da kommt halt der Status, wenn er eine Aktion durchgeführt hat. Aber es gibt AFAIK im custom Charger keinen 1p3p Befehl denman an den Go-E schicken könnte.

Was ich auch noch nicht ganz verstehe, ist das mit der Reihenfolge der Phasen. Bei Festinstallation kann man das ja machen. Wer aber die CE Stecker Version des Go-E hat (1 phasig blau/ 3 phasig rot), der weiß ja nie, auf welchen Phasen er in Relation zum Meter kommt. Eventuell habe ich das aber auch nur falsch verstanden. Ist aber für diesen PR auch nicht relevant.

premultiply commented:
> Was ich auch noch nicht ganz verstehe, ist das mit der Reihenfolge der Phasen. Bei Festinstallation kann man das ja machen. Wer aber die CE Stecker Version des Go-E hat (1 phasig blau/ 3 phasig rot), der weiß ja nie, auf welchen Phasen er in Relation zum Meter kommt. Eventuell habe ich das aber auch nur falsch verstanden.

Ja, das kann man beim go-E-Charger nicht falsch machen, da der interne Ladezähler zu Typ2-Anschluss eine feste und korrekte Zuordnung hat. Die netzseitige Beschaltungsreihenfolge spielt keine Rolle.

premultiply commented:
> Am liebsten hätte ich, wenn der Go-E über MQTT kommunizieren würde, weil da kommt halt der Status, wenn er eine Aktion durchgeführt hat. Aber es gibt AFAIK im custom Charger keinen 1p3p Befehl denman an den Go-E schicken könnte.

Bei der Kommunikationsschnittstelle sind wir prinzipiell völlig leidenschaftslos. In der Vergangenheit war es nur bei dem Teil leider so, dass nur die HTTP-Schnittstelle einigermaßen gepflegt und dokumentiert war. Bei Modbus oder MQTT haben dann immer relevante Features wie z. B. die Phasenumschaltung komplett gefehlt oder waren zumindest nicht dokumentiert.
Hat sich das denn inzwischen geändert?
